builtin/providers/ibmcloud/resource_ibmcloud_infra_virtual_guest.go

```diff
@@ -0,0 +1,1061 @@
+package ibmcloud
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"log"
+	"math"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/softlayer/softlayer-go/datatypes"
+	"github.com/softlayer/softlayer-go/filter"
+	"github.com/softlayer/softlayer-go/helpers/product"
+	"github.com/softlayer/softlayer-go/helpers/virtual"
+	"github.com/softlayer/softlayer-go/services"
+	"github.com/softlayer/softlayer-go/sl"
+)
+
+func resourceIBMCloudInfraVirtualGuest() *schema.Resource {
+	return &schema.Resource{
+		Create:   resourceIBMCloudInfraVirtualGuestCreate,
+		Read:     resourceIBMCloudInfraVirtualGuestRead,
+		Update:   resourceIBMCloudInfraVirtualGuestUpdate,
+		Delete:   resourceIBMCloudInfraVirtualGuestDelete,
+		Exists:   resourceIBMCloudInfraVirtualGuestExists,
+		Importer: &schema.ResourceImporter{},
+
+		Schema: map[string]*schema.Schema{
+			"hostname": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				DefaultFunc: genID,
+				DiffSuppressFunc: func(k, o, n string, d *schema.ResourceData) bool {
+					// FIXME: Work around another bug in terraform.
+					// When a default function is used with an optional property,
+					// terraform will always execute it on apply, even when the property
+					// already has a value in the state for it. This causes a false diff.
+					// Making the property Computed:true does not make a difference.
+					if strings.HasPrefix(o, "terraformed-") && strings.HasPrefix(n, "terraformed-") {
+						return true
+					}
+					return o == n
+				},
+			},
+
+			"domain": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"os_reference_code": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"image_id"},
+			},
+
+			"hourly_billing": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  true,
+				ForceNew: true,
+			},
+
+			"private_network_only": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+				ForceNew: true,
+			},
+
+			"datacenter": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"cores": {
+				Type:     schema.TypeInt,
+				Required: true,
+			},
+
+			"memory": {
+				Type:     schema.TypeInt,
+				Required: true,
+				ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+					memoryInMB := float64(v.(int))
+
+					// Validate memory to match gigs format
+					remaining := math.Mod(memoryInMB, 1024)
+					if remaining > 0 {
+						suggested := math.Ceil(memoryInMB/1024) * 1024
+						errors = append(errors, fmt.Errorf(
+							"Invalid 'ram' value %d megabytes, must be a multiple of 1024 (e.g. use %d)", int(memoryInMB), int(suggested)))
```

Comment:
The API server also responds back with memory in MB when there is read operation, so we simply put back that in the schema, that way it is more consistent.